### PR TITLE
Process directory files are cleaned up

### LINKED ISSUE #69: Cleanup up process files
I don't know how I forgot about this, but we're not cleaning up the `processDirPath` generated directories. We should - particularly given resource-constrained devices.

### PR DESCRIPTION
**Pull request by huntc**
We should always tidy up after ourselves.

Fixes #69

### CODE REVIEW

**Comment by dwijnand** [comment]
JFYI it seems deleting files on the JVM is a rabbit hole: https://github.com/sbt/io/issues/169. 🤷

**Comment by huntc** [comment]
Thanks @dwijnand - from having quickly read that issue, I'm thinking we should be ok as the JvmExecutor shouldn't be opening any files within a "process"'s directory structure.

**Comment by dwijnand** [comment]
It's hard to judge that code in a vacuum, as I suspect there are idiosyncrasies of sbt at play.

I just saw your code, remember the file visitor code that landed in sbt/io, and was trying to find you a better reference to use than a (seemingly) random gist.

In particular I'm not sure how wise it is to just swallow and ignore those two `exc: IOException`... I've also always started my file visitors extending `SimpleFileVisitor` for starting-point correctness.

**Comment by dwijnand** [comment]
On the other hand I don't want to fear-monger and yak-shave a `tidy-up` PR :wink:

**Comment by huntc** [comment]
Thanks again. I just poked around the SimpleFileVisitor. I think I'm comfortable swallowing those exceptions. My main goal here is to delete as many files as possible (hopefully all of them!). Disk space is a premium for where this is landing, so any savings there are going to be worth it.

Note that an IOException on the delete itself will cause the process to stop.

**Comment by huntc** [comment]
Actually, perhaps the IOException on the delete won't cause it to stop given the other CONTINUE results. However, I'm still happy if the function completes on a best-effort basis.

**Comment by dwijnand** [comment]
Sure, I just think it won't be very clear. I suspect it'll keep `FileVisitResult.CONTINUE`ing on every file failure and directory visit failure, and give you some unhelpful `IOException` about how it failed to delete the initial directory.

It will, though, have deleted everything else it possibly could in the process. So if that's the overarching priority it should work.

**Comment by huntc** [comment]
Thanks again for the feedback. I really appreciate it.

**Comment by huntc** [comment]
I just remembered. We should delete the process dir prior to running the process also. It is possible for an old one to be hanging around, particularly given restarts.

**Comment by longshorej** [comment]
I'm using something like this for our CQ tests -- the sorting should be enough to simplify the logic, right?

WRT files in use, I think that should only be a Windows issue; it's okay to unlink open files on POSIX. They'll be removed when the last file handle is closed.

```
  def deleteRecursively(path: java.nio.file.Path): Unit =
    Files
      .walk(path, java.nio.file.FileVisitOption.FOLLOW_LINKS)
      .sorted(java.util.Comparator.reverseOrder())
      .forEach(f => assert(f.toFile.delete()))
```

File: landlordd/daemon/src/main/scala/com/github/huntc/landlord/JvmExecutor.scala [hunk 1]
```diff
@@ -580,6 +604,8 @@ class JvmExecutor(
       context.stop(self)
   }
 
-  override def postStop(): Unit =
-    log.debug("Process actor stopped for {}", processId)
+  override def postStop(): Unit = {
+    log.debug("Process actor stopped for {} - removing its working dir at {}", processId, processDirPath)
+    deleteFiles(processDirPath)
```
Comment (longshorej) [inline landlordd/daemon/src/main/scala/com/github/huntc/landlord/JvmExecutor.scala]: Maybe submit this to the blocking dispatcher?

**Comment by huntc** [comment]
https://github.com/landlord/landlord/pull/70#issuecomment-404556120 was discussed during our standup. I quite like what we've got given that it considers an unknown number of files per directory. The streaming example could potentially exhaust memory in extreme cases.

**Comment by huntc** [comment]
Re https://github.com/landlord/landlord/pull/70#issuecomment-404493552

Deletion prior to running is now also considered.

**Pull request merged**